instrgen/docs/design.md

```diff
@@ -0,0 +1,185 @@
+# OpenTelemetry Go Source Automatic Instrumentation - design
+
+This document describes design of `instrgen` - compile time automatic instrumentation for GO programming language.
+
+## Background
+
+Usually techniques used for instrumenting programs in specific languages are connected to runtimes and type systems
+used by these languages and depend on capabilities offered by them. Methods differs depending on language properties.
+Is executed by virtual machine? Is statically or dynamically typed?
+What's matter is the injection point.
+In dynamic languages it's very easy to replace original implementation by providing a new one
+usually build on top of original one, a techique called monkey patching (look at python decorators).
+
+GO programming language is static, natively compiled programming language which means that works on bare methal without 
+virtual machine in between. The only component that lives between your program and machine is operating system.
+Therefore, to instrument GO programs, injection needs to be done at different level. There are two ways
+utilized for now by opentelemetry community for GO:
+
+1) an ebpf (https://ebpf.io/) a technology that allowes "programming" kernel 
+   and can be used by native programming languages in general on platforms that provide it.
+   
+2) compile time instrumentation - used by `instrgen`
+
+## Rationale
+
+Compile time instrumentation is not different by other mechanisms described above.
+The fundamental difference is the injection point, instrumentation has to be done 
+at compile time which has cons and pros.
+
+GO is perfect language for compile time instrumentation for two main reasons:
+- everything what's needed is in standard library
+- it's easy to control compilation process due to go compiler architecture
+
+## Design
+
+As mentioned above `instrgen` utilizes toolexec compiler switch (https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4058/).
+
+**NOTE** `driver` executable needs to be accessible via $PATH
+
+### Flow
+`instrgen` driver has two passes. First pass prepares some configuration for second pass based on provided settings,
```

Comment:
First pass is necessary to do some static analysis before providing data/suggestions needed for instance for UI. This can be also done as separate executable/program

Reply:
Is there a reason why someone would want to run the steps separately?